### PR TITLE
feat: labeler logic

### PR DESCRIPTION
**Pull request by niklastreml**
# Description
This pr implements a crossplane function, which allows any user to create a project scoped logging resource. It transforms a namespaced `logging.caas.telekom.de` resource into a clusterlevel `logging.banzaicloud.io` resource and ensure the created logging instance is scoped to the project in which the caas logging resource was created. For this two things need to be ensured:

1. spec.controlNamespace needs to be set to the namespace in which the namespaced caas logging resource was created. This ensure the FluentD deployment gets created in the namespace of the crd
2. spec.WatchNamespaces needs to point only to namespace inside that rancher project. This ensures that only logs from a users project are collected.

# Implementation
- Reads the namespace from a logging.caas.telekom.de CRD.
- Retrieves the value of the label `field.cattle.io/projectId` from that namespace.
- Creates a new banzaicloud logging resource.
- Sets its spec.controlNamespace to match the namespace of the logging.caas.telekom.de.
- Configures spec.watchNamespaceSelector.MatchLabels to {"field.cattle.io/projectId":"<usersProjectId>"}.

## Testing locally
Install the [crossplane cli](https://docs.crossplane.io/latest/cli/)

In one bash session start the function:
```bash
kubectl config use-context <your-rancher-cluster>
go run . --insecure --debug --out-of-cluster
```
This runs the function locally, using your current kubectl context to lookup the namespaces.

Invoke the function with a logging.caas.telekom.de Composite Resource:
```bash
cd examples/
crossplane beta render xr.yaml composition.yaml functions.yaml
```
```yaml
---
apiVersion: caas.telekom.de/v1alpha1
kind: Logging
metadata:
  name: example-xr
---
apiVersion: logging.banzaicloud.io/v1beta1
kind: Logging
metadata:
  annotations:
    crossplane.io/composition-resource-name: logging
  generateName: example-xr-
  labels:
    crossplane.io/composite: example-xr
  ownerReferences:
  - apiVersion: caas.telekom.de/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Logging
    name: example-xr
    uid: ""
spec:
  allowClusterResourcesFromAllNamespaces: false
  configCheck:
    timeoutSeconds: 0
  controlNamespace: microservices
  enableRecreateWorkloadOnImmutableFieldChange: false
  flowConfigCheckDisabled: false
  skipInvalidResources: false
  watchNamespaceSelector:
    matchLabels:
      field.cattle.io/projectId: caas-services
status:
  problemsCount: 0
```

This does not create any resources, it only invokes the function in the same way crossplane would, when actually deployed.
## Running Tests
```bash
go test ./...
```

# Example

I've deployed the function on my kind cluster. You can find an example that a user would create and its result below:
```yaml
# Input
apiVersion: caas.telekom.de/v1alpha1
kind: Logging
metadata:
  name: my-logging
  namespace: xrd-test
spec: {}
```
```yaml
# Output
apiVersion: logging.banzaicloud.io/v1beta1
kind: Logging
metadata:
  annotations:
    crossplane.io/composition-resource-name: logging
  creationTimestamp: "2024-02-08T15:21:25Z"
  generateName: my-logging-q66zj-
  generation: 1
  labels:
    crossplane.io/claim-name: my-logging
    crossplane.io/claim-namespace: xrd-test
    crossplane.io/composite: my-logging-q66zj
  name: my-logging-q66zj-5zssn
  ownerReferences:
  - apiVersion: caas.telekom.de/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: XLogging
    name: my-logging-q66zj
    uid: 0349b57d-79b3-4b08-8963-5a2eda90a5fd
  resourceVersion: "1406511"
  uid: fce5cc7d-9128-4f88-99ed-16b516bceb3e
spec:
  allowClusterResourcesFromAllNamespaces: false
  configCheck:
    timeoutSeconds: 0
  controlNamespace: xrd-test
  enableRecreateWorkloadOnImmutableFieldChange: false
  flowConfigCheckDisabled: false
  skipInvalidResources: false
  watchNamespaceSelector:
    matchLabels:
      field.cattle.io/projectId: p-12345
status:
  problems:
  - 'Deprecated behaviour! Other logging resources exist with the same loggingRef:
    hiller,kunde1,kunde2. This is going to be an error with the next major release.'
  problemsCount: 1
  watchNamespaces:
  - customer-1-ns-a
  - customer-1-ns-b
  - xrd-test
```

# External Resources
[Function Docs](https://docs.crossplane.io/latest/concepts/composition-functions/#how-composition-functions-work)
[Go example](https://docs.crossplane.io/knowledge-base/guides/write-a-composition-function-in-go/)

### CODE REVIEW

**Comment by eumel8** [comment]
LGTM, nice!

**Comment by niklastreml** [comment]
Ready for review

**Comment by CerRegulus** [comment]
LGTM. Thank you very much 🥳 ❤️

File: fn_test.go [hunk 1]
```diff
@@ -33,34 +35,123 @@ func TestRunFunction(t *testing.T) {
...
+								// 		Kind:       "Logging",
+								// 		APIVersion: "logging.banzaicloud.io/v1beta1",
+								// 	},
+								// 	Spec: v1beta1.LoggingSpec{
+								// 		ControlNamespace: "unit-test",
+								// 		WatchNamespaceSelector: &metav1.LabelSelector{
+								// 			MatchLabels: map[string]string{"field.cattle.io/projectId": "test-project"},
+								// 		},
+								// 	},
+								// }),
```
Comment (lvlcn-t) [inline fn_test.go]: Remove commented code

**Review by lvlcn-t — APPROVED** [review]
LGTM, you could move the logic into dedicated functions but for now it's readable.

File: fn.go [hunk 2]
```diff
@@ -5,36 +5,82 @@ import (
...
+		return rsp, nil
+	}
+
+	desired, err := request.GetDesiredComposedResources(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get desired composed resources"))
+		return rsp, nil
+	}
+
+	_ = v1beta1.AddToScheme(composed.Scheme)
```
Comment (puffitos) [inline fn.go]: Let's not ignore the error, you never know what could go wrong. It's go, we love boilerplate (`if err != nil` is our bread and butter!)

File: package/input/template.fn.crossplane.io_inputs.yaml [hunk 3]
```diff
@@ -26,19 +26,19 @@ spec:
...
-              you need. :)
-            type: string
           kind:
             description: 'Kind is a string value representing the REST resource this
               object represents. Servers may infer this from the endpoint the client
               submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
             type: string
           metadata:
             type: object
+          targetLabel:
```
Comment (puffitos) [inline package/input/template.fn.crossplane.io_inputs.yaml]: should we just name this `namespaceLabel` to be as precise as possible? That's what the label's content should be, right?

File: fn.go [hunk 4]
```diff
@@ -5,36 +5,82 @@ import (
...
+	l.Spec.WatchNamespaceSelector.MatchLabels = map[string]string{in.TargetLabel: projectid}
+	l.Spec.FluentdSpec = &v1beta1.FluentdSpec{}
+
+	cd, err := composed.From(l)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot convert %T to %T", l, &composed.Unstructured{}))
+		return rsp, nil
+	}
+
+	desired[resource.Name("logging")] = &resource.DesiredComposed{Resource: cd}
```
Comment (puffitos) [inline fn.go]: resource.Name is a `string`, no need for the cast.

**Review by puffitos — COMMENTED** [review]
LGTM, just a few remarks.

**Review by puffitos — APPROVED** [review]

**Pull request merged**